File: scripts/sign-did-configuration.py
#!/usr/bin/env python3
"""Sign a DID DomainLinkageCredential with W3C eddsa-jcs-2022."""
import argparse, hashlib, json, sys
from datetime import datetime, timezone
from pathlib import Path
import nacl.signing
from cryptography.hazmat.primitives.serialization import load_ssh_private_key

def load_key(path):
    from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
    key = load_ssh_private_key(Path(path).read_bytes(), password=None)
    if not isinstance(key, Ed25519PrivateKey): raise ValueError("Not Ed25519")
    return key.private_bytes_raw()

def _canonicalize(value):
    """RFC 8785 canonical JSON for this string-only credential profile."""
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"), sort_keys=True).encode()

def _base58(data):
    alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
    zeros = len(data) - len(data.lstrip(b"\0"))
    number = int.from_bytes(data, "big")
    encoded = ""
    while number:
        number, remainder = divmod(number, 58)
        encoded = alphabet[remainder] + encoded
    return "1" * zeros + encoded

def sign(cred, proof_options, seed):
    """W3C DI EdDSA hashing: SHA256(proof config) || SHA256(document)."""
    sk = nacl.signing.SigningKey(seed)
    proof_config = {**proof_options, "@context": cred["@context"]}
    hash_data = hashlib.sha256(_canonicalize(proof_config)).digest() + hashlib.sha256(_canonicalize(cred)).digest()
    return f"z{_base58(sk.sign(hash_data).signature)}"

def main():
    p = argparse.ArgumentParser()
    p.add_argument("--key")
    p.add_argument("--check", action="store_true")
    p.add_argument("--output", default="/root/ARIF-SITES/sites/arif-fazil.com/public/.well-known/did-configuration.json")
    args = p.parse_args()

    key_path = args.key
    if not key_path:
        for c in ["/root/.ssh/operator_did_ed25519", "/root/.secrets/vault-signing-ed25519"]:
            if Path(c).exists(): key_path = c; break
        if not key_path: print("ERROR: no key found", file=sys.stderr); sys.exit(1)

    print(f"Key: {key_path}")
    seed = load_key(key_path)
    sk = nacl.signing.SigningKey(seed)
    # Multikey uses the Ed25519 public-key multicodec prefix 0xed01.
    pub_mb = f"z{_base58(bytes.fromhex('ed01') + bytes(sk.verify_key))}"

    # Read expected key from DID document
    did_doc_path = Path("/root/ARIF-SITES/sites/arif-fazil.com/public/.well-known/did.json")
    did_doc = json.loads(did_doc_path.read_bytes())
    expected = did_doc["verificationMethod"][0]["publicKeyMultibase"]

    print(f"Public key:  {pub_mb}")
    print(f"DID expects: {expected}")
    match = pub_mb == expected
    print(f"Match: {match}")
    if args.check: sys.exit(0 if match else 1)
    if not match: print("ERROR: key mismatch"); sys.exit(1)

    credential = {
        "@context": ["https://www.w3.org/2018/credentials/v1", "https://w3id.org/security/multikey/v1"],
        "type": ["VerifiableCredential", "DomainLinkageCredential"],
        "issuer": "did:web:arif-fazil.com",
        "issuanceDate": "2026-04-30T00:00:00Z",
        "expirationDate": "2027-04-30T00:00:00Z",
        "credentialSubject": {"id": "did:web:arif-fazil.com", "origin": "https://arif-fazil.com"},
    }
    now = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
    proof = {
        "type": "DataIntegrityProof", "cryptosuite": "eddsa-jcs-2022",
        "created": now, "verificationMethod": f"did:web:arif-fazil.com#{did_doc['verificationMethod'][0]['id'].split('#')[-1]}",
        "proofPurpose": "assertionMethod",
    }
    proof["proofValue"] = sign(credential, proof, seed)
    did_config = {
        "@context": "https://identity.foundation/well-known/did-configuration/v1",
        "linked_dids": [{
            **credential,
            "proof": proof,
        }],
    }
    out = Path(args.output)
    out.parent.mkdir(parents=True, exist_ok=True)
    out.write_text(json.dumps(did_config, indent=2) + "\n")
    print(f"\nSigned: {out}")

    deploy = Path("/var/www/html/arif/.well-known/did-configuration.json")
    if deploy.parent.exists():
        deploy.write_text(json.dumps(did_config, indent=2) + "\n")
        print(f"Deployed: {deploy}")

    # Also deploy updated did.json
    did_deploy = Path("/var/www/html/arif/.well-known/did.json")
    if did_deploy.parent.exists():
        did_deploy.write_bytes(did_doc_path.read_bytes())
        print(f"Deployed: {did_deploy}")

if __name__ == "__main__":
    main()

File: scripts/sign-did-configuration.mjs
#!/usr/bin/env node
import { createHash, sign } from "node:crypto";
import { readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { homedir } from "node:os";

const BASE58_ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

function canonicalJson(value) {
  if (Array.isArray(value)) {
    return `[${value.map(canonicalJson).join(",")}]`;
  }
  if (value && typeof value === "object") {
    return `{${Object.keys(value)
      .sort()
      .map((key) => `${JSON.stringify(key)}:${canonicalJson(value[key])}`)
      .join(",")}}`;
  }
  return JSON.stringify(value);
}

function base58btcEncode(buffer) {
  let carry;
  let digits = [0];
  for (let i = 0; i < buffer.length; i++) {
    carry = buffer[i];
    for (let j = 0; j < digits.length; j++) {
      carry += digits[j] << 8;
      digits[j] = carry % 58;
      carry = Math.floor(carry / 58);
    }
    while (carry > 0) {
      digits.push(carry % 58);
      carry = Math.floor(carry / 58);
    }
  }
  for (let i = 0; i < buffer.length; i++) {
    if (buffer[i] !== 0) break;
    digits.push(0);
  }
  return "z" + digits.reverse().map((d) => BASE58_ALPHABET[d]).join("");
}

function argValue(flag, fallback) {
  const index = process.argv.indexOf(flag);
  if (index === -1) return fallback;
  return process.argv[index + 1] || fallback;
}

function main() {
  const defaultKeyPath = join(homedir(), ".ssh", "id_ed25519");
  const keyPath = argValue("--key", defaultKeyPath);
  const configPath = argValue("--config", "./sites/arif-fazil.com/public/.well-known/did-configuration.json");

  console.log(`[+] Loading did-configuration from: ${configPath}`);
  let configData;
  try {
    configData = JSON.parse(readFileSync(configPath, "utf8"));
  } catch (error) {
    console.error(`[-] Failed to read config: ${error.message}`);
    process.exit(1);
  }

  const credential = configData.linked_dids?.[0]?.credentials?.[0];
  if (!credential) {
    console.error("[-] No domain linkage credential structure found in config.");
    process.exit(1);
  }

  // 1. Prepare unsigned credential payload
  const unsignedCredential = { ...credential };
  delete unsignedCredential.proof;

  // Include top-level fields required for validation context
  unsignedCredential.issuanceDate = configData.linked_dids[0].issuance_date;
  unsignedCredential.issuer = configData.linked_dids[0].issuer;

  const canonicalPayload = canonicalJson(unsignedCredential);
  console.log(`[+] Canonical Payload length: ${canonicalPayload.length} bytes`);

  // 2. Load private key
  console.log(`[+] Loading private key from: ${keyPath}`);
  let privateKey;
  try {
    privateKey = readFileSync(keyPath, "utf8");
  } catch (error) {
    console.error(`[-] Failed to read private key: ${error.message}`);
    console.log("[!] Please specify your private key path with: --key <path_to_private_key>");
    process.exit(1);
  }

  // 3. Sign the payload
  console.log("[+] Signing payload using Ed25519...");
  let signatureBuffer;
  try {
    signatureBuffer = sign(null, Buffer.from(canonicalPayload), privateKey);
  } catch (error) {
    console.error(`[-] Signing failed: ${error.message}`);
    process.exit(1);
  }

  // 4. Encode and insert
  const signatureBase58 = base58btcEncode(signatureBuffer);
  console.log(`[+] Encoded signature: ${signatureBase58}`);

  credential.proof.proofValue = signatureBase58;
  // Make sure the verificationMethod is correct
  credential.proof.verificationMethod = "did:web:arif-fazil.com#arif-fazil";

  // 5. Save back
  writeFileSync(configPath, JSON.stringify(configData, null, 2) + "\n");
  console.log(`[+] SUCCESS — updated did-configuration.json with real signature.`);
}

main();
